### PR TITLE
Use `Serializer.equals` when comparing parameters for cache reuse 

### PR DESCRIPTION
**Pull request by lacasseio**
This PR address the concern for the short term fix of gradle/gradle#933. It implements the `equals` and `hashCode` for all `Serializer` implementation. We use this equality to check for matching cache parameter and determine if a cache instance can be reused. This patch builds on https://github.com/gradle/gradle/pull/1011 which is why it target branch `dl-issue-933`. Once the code review is done, the change will be manually merged.

### Open issues
- There is an ongoing discussion with @big-guy to see if we could ignore some `Serializer` in this PR and achieve the same level of robustness.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java [hunk 1]
```diff
@@ -573,9 +573,9 @@ private void checkCompatibleValueSerializer(Collection<String> faultMessages, Se
...
 
         private void checkCompatibleCacheDecorator(Collection<String> faultMessages, CacheDecorator cacheDecorator) {
-            String requestClassName = ClassUtils.getShortCanonicalName(cacheDecorator, null);
-            String currentClassName = ClassUtils.getShortCanonicalName(parameters.getCacheDecorator(), null);
-            if (!Objects.equal(requestClassName, currentClassName)) {
+            if (!Objects.equal(cacheDecorator, parameters.getCacheDecorator())) {
+                String requestClassName = ClassUtils.getShortCanonicalName(cacheDecorator, null);
+                String currentClassName = ClassUtils.getShortCanonicalName(parameters.getCacheDecorator(), null);
                 faultMessages.add(
                     String.format(" * Requested cache decorator type (%s) doesn't match current cache type (%s)",
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java]: I think it's OK to just rely on the default `toString` for the two cache decorators.  It'll work if the cache decorators are the same type, but different instances or if they are different types or if only one cache decorator is null.

If we just show class names, you won't be able to tell there are two different instances of the same class.

File: subprojects/messaging/src/main/java/org/gradle/internal/serialize/AbstractSerializer.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+package org.gradle.internal.serialize;
+
+import com.google.common.base.Objects;
+
+/**
+ * This abstract class provide a sensible default implementation for {@code Serializer} equality. This equality
+ * implementation is required to enable cache instance reuse within the same Gradle runtime. Serializers are used
+ * as cache parameter which need to be compared to determine compatible cache.
+ *
+ * \see https://github.com/gradle/gradle/issues/933
```
Comment (big-guy) [inline subprojects/messaging/src/main/java/org/gradle/internal/serialize/AbstractSerializer.java]: You have to use `@see <a href="http://url">text</a>` for it to render properly in JavaDoc.

I think it's OK to just include the issue number in the commit messages and in a `@Issue(url)` in the tests (if you have tests dedicated to this particular bug/issue).

**Comment by lacasseio** [comment]
@big-guy I fixed what you commented on. I also tried to narrow the change set a bit more. These `Serializer` are the one that seems to be directly used in a cache file. I could narrow it down even more but we may overlook a `Serializer` and have the same problem later. Have a look and tell me if you want me to make any additional verification.

**Comment by lacasseio** [comment]
Sorry for spamming so many people during a bad assignment of the task to myself. My browser froze up and when it came back a gazillion people was selected. Feel free to use the in-mail unsubscribe button to unfollow this thread. Again, truly sorry for the inconvenient.

**Comment by big-guy** [comment]
This LGTM @lacasseio let's try merging it.

**Comment by adammurdoch** [comment]
yes, something like that. I was thinking we would just use a field in the various places where `createCache()` are called.

**Comment by big-guy** [comment]
OK.  

To use the same instance of `TaskExecutionListSerializer` for all `CacheBackedTaskHistoryRepository` instances (this is created in `TaskExecutionServices`), we'd have to 

1) inject it from a service registry shared across `GradleBuild`s (so probably in global services) or
2) make it a static field (harder since this needs a `StringInterner` too) or 
3) have a factory that always returns the same instance of  `TaskExecutionListSerializer`. 

I think my preference would be with the global services registry (and expose something like `BaseSerializerFactory` and `SerializerRegistry`), so we'd know where all the serializers were coming from.  Are any of the other options more preferable? 

Ugh.. looking at `TaskExecutionListSerializer` more now, I think it may not be multi-build safe.  I don't see how the `AsyncCacheAccessContext` magic works since the same `CONTEXT_KEY_FOR_CLASSLOADER` would be created for all builds (that could have very different classloaders).

**Comment by adammurdoch** [comment]
We should make it a service. The `AsyncCacheAccessContext` thing can go once we no longer store the actual property values in the history object, but instead store some serialized or hashed form of the values.

**Comment by big-guy** [comment]
I pushed some changes to https://github.com/gradle/gradle/tree/sg-serializer-service, @lacasseio do you have a sample build you've been testing for cache corruption problems?

**Comment by lacasseio** [comment]
Thanks @big-guy for helping out on this PR. You should be able to reproduce the corruption with the following steps:
1. Clone apache/groovy
2. Checkout `0e51066afeaae7d8b862a7dccf971c2bdcae0541` (older commit on `parrot` branch)
3. Execute `gradlew -PuseAntlr4=true clean installGroovy`
4. Execute `gradlew -PuseAntlr4=true clean installGroovy` a second time. You should see a corruption detection from Gradle where it will discard the `fileSnapshots.bin` cache and fail when retrieving something from `taskArtifacts.bin`.

Don't hesitate to ping me if you are having a hard time reproducing the corruption.

**Comment by adammurdoch** [comment]
@big-guy the changes look good to me

**Pull request merged**